stdlib/runtime/runtimeDebug.gr

```diff
@@ -0,0 +1,712 @@
+          "; new count: " ++
+          toString(newRefCount) ++
+          "; tag: " ++
+          toString(tag) ++
+          " " ++
+          tagToTypeName(tag) ++
+          ") " ++
+          maybeObjectContentsAppendix(userPtr),
+        )
+      }
+    }
+  }
+
+  let onDecRef = (userPtr: Number, ignoreZeros: Bool) => {
+    totalDecRefs += 1
+
+    let oldRefCount = readRefCount(userPtr)
+    // It would be more accurate to actually receive the new ref count,
+    // but it would complicate things unnecessarily. Instead I'm replicating
+    // a piece of logic from GC.decRef regarding decrementing objects with
+    // refcount of 0.
+    let newRefCount = if (oldRefCount == 0 && ignoreZeros) {
+      0
+    } else {
+      oldRefCount - 1
+    }
+    let tag = readTag(userPtr)
+
+    incrementCountInMap(tag, decRefsByTag)
+
+    if (options.printIncrefsAndDecRefs) {
+      if (oldRefCount < 1 && !ignoreZeros) {
+        print(
+          "⚠ PROBLEM DETECTED: unexpected decRef at address 0x" ++
+          " (tag: " ++
+          toString(tag) ++
+          " " ++
+          tagToTypeName(tag) ++
+          ")" ++
+          maybeObjectContentsAppendix(userPtr),
+        )
+
+        if (tag == 7
+        /* or Tags._GRAIN_LAMBDA_HEAP_TAG*/) {
+          print(
+            "The object is a function or lambda. You have to incRef a function before calling it from a @disableGC context.",
+          )
+        }
+      } else {
+        if (Set.contains(userPtr, allocations)) {
+          print(
+            "decref for object at 0x" ++ toHex(userPtr) ++ " (prev count: " ++
+            toString(oldRefCount) ++
+            "; new count: " ++
+            toString(newRefCount) ++
+            "; tag: " ++
+            toString(tag) ++
+            " " ++
+            tagToTypeName(tag) ++
+            ")" ++
+            maybeObjectContentsAppendix(userPtr),
+          )
+        } else {
+          if (!Map.contains(userPtr, inintialRefCountsOnPreExistingObjects)) {
+            Map.set(userPtr, oldRefCount, inintialRefCountsOnPreExistingObjects)
+          }
+
+          incrementCountInMap(userPtr, decRefsOnPreExistingObjects)
+
+          print(
+            "decref for pre-existing object at 0x" ++ toHex(userPtr) ++
+            " (prev count: " ++
+            toString(oldRefCount) ++
+            "; new count: " ++
+            toString(newRefCount) ++
+            "; tag: " ++
+            toString(tag) ++
+            " " ++
+            tagToTypeName(tag) ++
+            ")" ++
+            maybeObjectContentsAppendix(userPtr),
+          )
+        }
+      }
+    }
+  }
+
+  let onMalloc = (userPtr: Number) => {
+    totalMallocs += 1
+
+    // Tracking objects allocated in the debug block.
+    if (Set.contains(userPtr, allocations)) {
+      // A new allocation, but for an object already allocated inside the debug block, but never removed.
+      print(
+        "⚠ PROBLEM DETECTED: unexpected malloc at address 0x" ++
```

Comment:
Sort of wish we had colored output here hehe
(not _really_ requesting this, but just a thought)